File: oracle-js/ggl_lower.js
export class LowerError extends Error {
  constructor(code, msg) {
    super(msg);
    this.code = code;
    this.msg = msg;
  }
}

export function lowerAstToSceneXjson(ast, grammarAbi) {
  if (!ast || typeof ast !== "object") {
    throw new LowerError("E_LOWER_AST", "invalid AST");
  }
  return {
    "@scene": "asx://schema/scene.ir.v1",
    grammar: grammarAbi?.id,
    body: ast,
  };
}

File: cline/GglLower.java
import java.util.LinkedHashMap;
import java.util.Map;

public final class GglLower {
    public static final class LowerError extends RuntimeException {
        public final String code;

        public LowerError(String code, String msg) {
            super(msg);
            this.code = code;
        }
    }

    public static Map<String, Object> lower(Object ast, Object grammarAbi) {
        if (!(ast instanceof Map)) {
            throw new LowerError("E_LOWER_AST", "invalid AST");
        }
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("@scene", "asx://schema/scene.ir.v1");
        out.put("grammar", grammarAbi);
        out.put("body", ast);
        return out;
    }
}
